packages/contracts/contracts/L1/L2OutputOracle.sol

```diff
@@ -0,0 +1,120 @@
+
+    // The interval in seconds at which checkpoints must be submitted.
+    uint256 public immutable submissionInterval;
+
+    // The time between blocks on L2.
+    uint256 public immutable l2BlockTime;
+
+    // The number of blocks in the chain before the first block in this contract.
+    uint256 public immutable historicalTotalBlocks;
+
+    // The timestamp of the first L2 block recorded in this contract.
+    uint256 public immutable startingBlockTimestamp;
+
+    // The timestamp of the most recent L2 block recorded in this contract.
+    uint256 public latestBlockTimestamp;
+
+    // A mapping from L2 timestamps to the output root for the block with that timestamp
+    mapping(uint256 => bytes32) public l2Outputs;
+
+    /**
+     * Initialize the L2OutputOracle contract.
+     * @param _submissionInterval The desired interval in seconds at which
+     *        checkpoints must be submitted.
+     * @param _l2BlockTime The desired L2 inter-block time in seconds.
+     * @param _genesisL2Output The initial L2 output of the L2 chain.
+     * @param _historicalTotalBlocks The number of blocks that preceding the
+     *        initialization of the L2 chain.
+     */
+    constructor(
+        uint256 _submissionInterval,
+        uint256 _l2BlockTime,
+        bytes32 _genesisL2Output,
+        uint256 _historicalTotalBlocks,
+        address sequencer
+    ) {
+        submissionInterval = _submissionInterval;
+        l2BlockTime = _l2BlockTime;
+        l2Outputs[block.timestamp] = _genesisL2Output; // solhint-disable not-rely-on-time
+        historicalTotalBlocks = _historicalTotalBlocks;
+        latestBlockTimestamp = block.timestamp; // solhint-disable not-rely-on-time
+        startingBlockTimestamp = block.timestamp; // solhint-disable not-rely-on-time
+
+        _transferOwnership(sequencer);
+    }
+
+    /**
+     * Accepts an L2 output checkpoint and the timestamp of the corresponding L2
+     * block. The timestamp must be equal to the current value returned by
+     * `nextTimestamp()` in order to be accepted.
+     * This function may only be called by the Sequencer.
+     * @param _l2Output The L2 output of the checkpoint block.
+     * @param _l2timestamp The L2 block timestamp that resulted in _l2Output.
+     * @param _blockhash A block hash which must be included in the current chain.
+     * @param _blocknumber The block number with the specified block hash.
+     */
+    function appendL2Output(
+        bytes32 _l2Output,
+        uint256 _l2timestamp,
+        bytes32 _blockhash,
+        uint256 _blocknumber
+    ) external payable onlyOwner {
+        require(_l2timestamp < block.timestamp, "Cannot append L2 output in future");
+        require(_l2timestamp == nextTimestamp(), "Timestamp not equal to next expected timestamp");
+        require(_l2Output != bytes32(0), "Cannot submit empty L2 output");
+
+        if (_blockhash != bytes32(0)) {
```

Comment:
If I understand this correctly, its to protect against time bandit attacks?

Reply:
Yes - the `_blockhash` prevents the tx from being replayed in case of a L1 reorg (re-orgin past `_blocknumber`), which would allow a successful fault proof agains the now-incorrect output root.

The blockhash can only be check for ~256 blocks though, so L1 validators might censor for the duration. In that case we can omit the block hash and wait long enough we are sure the block at `_blocknumber` won't be re-orged, in which case replays won't do any damage.